feat(native/doc): first understanding (#430)
citizenfx/fivem-docs: content/docs/scripting-manual/introduction/about-native-functions.md

---

> title: Understanding and Using Native Functions in FiveM

Reviewer: This document shouldn't be game specific they work the same across different games (GTA5, RDR2, GTA4)

Maybe something like `Understanding and Using Natives on supported platforms`

Author: It's only for `FiveM` - spoke to intern people about this.

weight: 1300
---

> In FiveM, you have access to what are called "native functions" or simply "natives". These functions interact with the game directly, allowing you to perform a wide variety of actions within the game environment. This guide will explain what natives are, where you can find them, and how to use them in your scripts.

Reviewer: Same note as above

Rewording could be:
`In the games supported by the platform,`

Author: Same comment as above

## What Are Native Functions?
Native functions in FiveM are predefined functions that enable interaction with the game engine. They are essential for creating and managing game mechanics, player interactions, and various other functionalities within your server. Natives are used extensively in both client-side and server-side scripts to control gameplay elements, manipulate objects, manage player states, and much more.

> These functions are called "native" because they are built into the game engine itself. They provide a powerful interface for developers to interact with the underlying game code. Natives can perform a wide range of tasks, from simple operations like printing messages to the console, to complex actions such as manipulating player physics or creating game entities.

Reviewer 2: I'd rephrase this since natives are not only implemented in the game itself, the server also implements the natives system.

## Where Can You Find Native Functions?

> FiveM provides comprehensive documentation for native functions, which can be found on our official website. This documentation is essential for understanding what each native function does, how to use it, and what parameters it requires.

Reviewer: `FiveM` -> `Cfx`

Author: Same comment as above

> - **FiveM Native Reference**: The primary resource for native functions is the [FiveM Native Reference](https://docs.fivem.net/natives/). This reference includes a searchable database of all available native functions, along with detailed descriptions, parameter lists, and usage examples.

Reviewer: Might want to include RedM native references too even if they're unofficial 

https://redm.disquse.me/natives/

Author: They want to include only official links to the docs even if Disquse is responsible for RedM.

> - **FiveM Community Forums**: The [FiveM forums](https://forum.fivem.net/) are another valuable resource. Here, you can find discussions, tutorials, and examples from other developers who share their experiences and solutions related to using native functions.

Reviewer: This should be just `Community Forums` and `[Cfx.re Forums](https://forum.cfx.re/)`

Author: I agree yeah

> - **CitizenFX GitHub**: Users can contribute to the [CitizenFX GitHub repository](https://github.com/citizenfx/fivem) by adding descriptions, names, and new native functions. This collaborative effort helps keep the list of natives comprehensive and up-to-date, ensuring that the community benefits from the latest features and improvements.

Reviewer 3: Maybe it would be worth including this repo too ?
https://github.com/citizenfx/natives

> ## Difference Between Client Natives and Server-Side Natives

Reviewer 3: It should either be "Client-Side" and "Server-Side" or "Client" and "Server"

> ### Client Natives

Reviewer: It might be worth noting here `Cfx Natives` as they are used when a resource uses `game "common"`

Client natives are functions that run on the player's game client. These functions are primarily used to handle tasks related to the player's immediate environment, rendering, input handling, and interactions that need to occur on the client side. Examples include manipulating the player’s view, handling UI elements, and interacting with client-specific data.

> ### Server-Side Natives

Reviewer 3: Same here, you have to be consistent with naming, if you put "side" then do the same for client

> Server-side natives, on the other hand, run on the server. These functions manage game logic that needs to be controlled and validated by the server, such as player authentication, database interactions, and ensuring game state consistency. Server-side natives are crucial for maintaining the integrity and security of the game.

Reviewer 3: Natives don't allow database interaction AFAIK ?

1. **Create a Server Script**: In your resource folder, create a new server script file, e.g., `main_sv.lua`.
2. **Register a Command and Use a Native Function**: Use the native function `GetPlayerName` to get the name of a player and print it to the server console.

```lua
-- main_sv.lua

-- Register a command to get the player name
RegisterCommand('getplayername', function(source, args, rawCommand)
    local playerName = GetPlayerName(source)
    print('Player Name: ' .. playerName)
end, false)
```

### Create Your First Resource
To learn how to create your first resource step-by-step, refer to the comprehensive guide in [this section](/content/docs/scripting-manual/introduction/creating-your-first-script-extra.md). This guide will walk you through the process of setting up a new resource, writing your first script, and integrating it into your FiveM server.

### Commonly Used Native Functions
Here are some commonly used native functions and their descriptions:

- **AddTextEntry**: Adds a text entry to the game.
  ```lua
  AddTextEntry(key, value)
  ```

- **GetPlayerName**: Returns the name of a player.
  ```lua
  local playerName = GetPlayerName(playerId)
  ```

> - **DrawNotification**: Displays a notification on the screen.

Reviewer 3: This is not a native anymore (an old name) and I think it's not good practice to include old names in the official docs ! New name is `EndTextCommandThefeedPostTicker`.
Also this is not one of the easiest native to include in a "first understanding" docs imo. Maybe include like `GetEntityCoords` or `DeleteEntity`

These functions are called "native" because they are built into the game engine itself (for most of them). Like stated, some of them are made CFX to handle particular things in the platform. Natives provide a powerful interface for developers to interact with the game code. Natives can perform a wide range of tasks, from simple operations like getting a player name, to complex actions such as manipulating player physics or creating game entities.

## Where Can You Find Native Functions?
FiveM provides comprehensive documentation for native functions. This documentation is essential for understanding what each native function does, how to use it, and what parameters it requires.

- **FiveM Natives**: All documented natives can be found on [our official website](https://docs.fivem.net/natives/).

- **Community Forums**: The [Cfx.re Forums](https://forum.cfx.re/) are another valuable resource. Here, you can find discussions, tutorials, and examples from other developers who share their experiences and solutions related to using native functions.

> - **CitizenFX GitHub**: Users can contribute to the [repository](https://github.com/citizenfx/fivem) by adding descriptions, names, and new native functions. This collaborative effort helps keep the list of natives comprehensive and up-to-date, ensuring that the community benefits from the latest features and improvements.

Reviewer: ```suggestion
- **Native Docs Repo**: Users can contribute to the [repository](https://github.com/citizenfx/natives) by adding descriptions, names, and new native functions.
```
